src/general/ShuffleBag.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+
+public class ShuffleBag<T> : IEnumerable<T>
+{
+    private Random random;
+    private List<T> initialContent;
+    private List<T> currentContent;
+
+    public ShuffleBag(Random random)
+    {
+        initialContent = new List<T>();
+        currentContent = new List<T>();
+        this.random = random;
+    }
+
+    public void Clear()
+    {
+        initialContent.Clear();
+        currentContent.Clear();
+    }
+
+    public void Add(T element)
+    {
+        initialContent.Add(element);
+        currentContent.Add(element);
+    }
+
+    public bool Remove(T element)
+    {
+        currentContent.Remove(element);
+        return initialContent.Remove(element);
+    }
+
+    public void RemoveAll(Predicate<T> f)
+    {
+        initialContent.RemoveAll(f);
+    }
+
+    public void FillAndShuffle()
+    {
+        currentContent.Clear();
+
+        foreach (var element in initialContent)
+            currentContent.Add(element);
+
+        Shuffle();
+    }
+
+    public T PickAndDrop()
+    {
+        var drawnElement = Pick();
+        Drop();
+
+        return drawnElement;
+    }
+
+    public IEnumerator<T> GetEnumerator()
+    {
+        return new ShuffleBagEnumerator<T>(this);
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+
+    /// <summary>
+    ///   Fisher–Yates Shuffle Alg. (https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle)
+    ///   Provides a uniform random shuffle in O(n)
+    /// </summary>
+    private void Shuffle()
+    {
+        for (int i = 0; i < currentContent.Count - 2; i++)
+        {
+            int j = random.Next(i, currentContent.Count);
+
+            T swapedElement = currentContent[i];
+            currentContent[i] = currentContent[j];
+            currentContent[j] = swapedElement;
+        }
+    }
+
+    private T Pick()
+    {
+        if (currentContent.Count == 0)
+        {
+            FillAndShuffle();
+        }
+
+        return currentContent[currentContent.Count - 1];
+    }
+
+    private bool Drop()
+    {
+        if (currentContent.Count == 0)
+            throw new NullReferenceException("Cannot drop from empty bag!");
```

Comment:
use an `InvalidOperationException` here. It isn't really a null reference problem.